Extend Functionality of ESPHome Media Player

rwrozelle opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
No breaking change

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
# What does this implement/fix?
Lays the ground work for child components of Esphome Media Player to be built with a richer set of capability.

## Description
Extends Esphome Media Player component to support the following commands in child components:
* Next Track
* Previous Track
* Turn On
* Turn Off
* Clear Playlist
* Shuffle
* Unshuffle
* Repeat Off
* Repeat One
* Repeat All
* Join
* Unjoin

Traits are extended to include:
* supports next/previous track
* supports turn off/on
* supports grouping

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

Requires the following pull requests:
https://github.com/esphome/esphome-docs/pull/4090
https://github.com/esphome/aioesphomeapi/pull/911
https://github.com/esphome/esphome/pull/7158

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:
Nothing changes for existing esphome media_players, new functionality requires updated/new esphome components.
An example of which is my external component https://github.com/rwrozelle/audio-media-player
- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]
Documentation updated on esphome.io repository.

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

rwrozelle commented:
> Hi @bobrozelle
> 
> It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant/core#122799).
> 
> Once you do that we will be able to review and accept this pull request.
> 
> Thanks!

When I try, I get this error "No pending request found for rwrozelle. Are you signing the CLA with the same GitHub user that created the commits in the Pull Request?"

bdraco commented:
It looks like you are committing from two different accounts.  Please rework the commits so they come from an account that has signed the CLA

rwrozelle commented:
Thank you, its obvious now that you've pointed it out.  I've signed cla with both accounts.

rwrozelle commented:
I'll work on getting the files updated for ruff.  A little confused because it passed ruff in my local dev environment.  Most of the MyPY errors look like they are because of the changes needed from https://github.com/esphome/aioesphomeapi/pull/911.  How is that handled?

rwrozelle commented:
I've updated the code and updated for all pre-commit requested changes from my local dev environment.  I'm going to submit for review, but still unclear if it will pass tests because of changes needed from https://github.com/esphome/aioesphomeapi/pull/911.  Is there a way to get the github tester to use updates from another pending Pull Request?
